src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1558,30 +1617,32 @@ private bool TryProcessCompilationUnitCompleted(CompilationUnitCompletedEvent co
                 return true;
             }
 
+            var processedAnalyzers = analysisStateOpt != null ? PooledHashSet<DiagnosticAnalyzer>.GetInstance() : null;
             try
             {
                 var success = true;
-                foreach (var analyzer in analysisScope.Analyzers)
+                foreach (var (analyzer, semanticModelActions) in _semanticModelActions)
                 {
-                    ImmutableArray<SemanticModelAnalyzerAction> semanticModelActions;
-                    if (_semanticModelActionsMap.TryGetValue(analyzer, out semanticModelActions))
+                    if (!analysisScope.Contains(analyzer))
                     {
-                        // Execute actions for a given analyzer sequentially.
-                        if (!AnalyzerExecutor.TryExecuteSemanticModelActions(semanticModelActions, analyzer, semanticModel, completedEvent, analysisScope, analysisStateOpt, isGeneratedCode))
-                        {
-                            success = false;
-                        }
+                        continue;
                     }
-                    else
+
+                    // Execute actions for a given analyzer sequentially.
+                    if (!AnalyzerExecutor.TryExecuteSemanticModelActions(semanticModelActions, analyzer, semanticModel, completedEvent, analysisScope, analysisStateOpt, isGeneratedCode))
                     {
-                        analysisStateOpt?.MarkEventComplete(completedEvent, analyzer);
+                        success = false;
                     }
+
+                    processedAnalyzers?.Add(analyzer);
                 }
 
+                analysisStateOpt?.MarkEventCompleteForUnprocessedAnalyzers(completedEvent, analysisScope, processedAnalyzers);
                 return success;
             }
             finally
             {
+                processedAnalyzers?.Free();
                 completedEvent.FlushCache();
             }
         }
```

Comment:
fwiw, this repeated pattern seems unfortunate.  have you considerd a helper class that has the general pattern in teh abstract base, and derived types supply the code to do the different stuff like `TryExecuteSemanticModelActions`.

Reply:
or a helper function with teh core logic that is passed a delegate to execute to do the specialized work.

Reply:
As we discussed offline, there is more then one place where a different API is being invoked with different parameters, and we want to avoid delegate allocations here as this is a hot path. We can potentially refactor to have separate base/derived types, but that seems too heavyweight for sharing just few lines of code. I am going to stay with the current approach for now and refactor in future when it seems there is a bigger value add in such a refactoring.